src/csharp/Microsoft.Spark/Sql/Functions.cs

```diff
@@ -4091,6 +4279,22 @@ private static UserDefinedFunction CreateVectorUdf<TResult>(string name, Delegat
                 UdfUtils.GetReturnType(typeof(TResult)));
         }
 
+        private static UserDefinedFunction CreateUdfWithReturnType<TResult>(
+            string name,
+            Delegate execute,
+            UdfUtils.PythonEvalType evalType,
+            StructType returnType)
+        {
+            return UserDefinedFunction.Create(
+                name,
+                CommandSerDe.Serialize(
+                    execute,
+                    CommandSerDe.SerializedMode.Row,
+                    CommandSerDe.SerializedMode.Row),
+                evalType,
+                returnType.Json);
+        }
+
```

Comment:
How about combining to something like:
```csharp
        private static UserDefinedFunction CreateUdf<TResult>(
            string name,
            Delegate execute,
            UdfUtils.PythonEvalType evalType) =>
            CreateUdf(name, execute, evalType, UdfUtils.GetReturnType(typeof(TResult)));

        private static UserDefinedFunction CreateUdf(
            string name,
            Delegate execute,
            UdfUtils.PythonEvalType evalType,
            StructType returnType) =>
            CreateUdf(name, execute, evalType, returnType.Json);

        private static UserDefinedFunction CreateUdf(
            string name,
            Delegate execute,
            UdfUtils.PythonEvalType evalType,
            string returnType) => UserDefinedFunction.Create(name,
                CommandSerDe.Serialize(
                    execute,
                    CommandSerDe.SerializedMode.Row,
                    CommandSerDe.SerializedMode.Row),
                evalType, returnType);
```

Then you can also change
- 1
    ```csharp
            public static Func<Column> Udf(Func<GenericRow> udf, StructType returnType)
            {
                return CreateUdf<GenericRow>(udf.Method.ToString(), UdfUtils.CreateUdfWrapper(udf), returnType).Apply0;
            }
    ```
    to
    ```csharp
            public static Func<Column> Udf(Func<Row> udf, StructType returnType)
            {
                return CreateUdf(udf.Method.ToString(), UdfUtils.CreateUdfWrapper(udf), returnType).Apply0;
            }
    ```

and
- 2 
    ```csharp
            private static UserDefinedFunction CreateUdf<TResult>(string name, Delegate execute, StructType returnType)
            {
                return CreateUdfWithReturnType<TResult>(name, execute, UdfUtils.PythonEvalType.SQL_BATCHED_UDF, returnType);
            }
    ```
    to
    ```csharp
            private static UserDefinedFunction CreateUdf(string name, Delegate execute, StructType returnType)
            {
                return CreateUdf(name, execute, UdfUtils.PythonEvalType.SQL_BATCHED_UDF, returnType);
            }
    ```

Reply:
> Then you can also change
> 
> * 1
>   ```cs
>           public static Func<Column> Udf(Func<GenericRow> udf, StructType returnType)
>           {
>               return CreateUdf<GenericRow>(udf.Method.ToString(), UdfUtils.CreateUdfWrapper(udf), returnType).Apply0;
>           }
>   ```
>   
>   
>   to
>   ```cs
>           public static Func<Column> Udf(Func<Row> udf, StructType returnType)
>           {
>               return CreateUdf(udf.Method.ToString(), UdfUtils.CreateUdfWrapper(udf), returnType).Apply0;
>           }
>   ```
> 
Thanks for your suggestion. In the to function part, do you mean `Func<Row>` or Func<GenericRow> as below?
```cs
public static Func<Column> Udf(Func<GenericRow> udf, StructType returnType)
{
    return CreateUdf(udf.Method.ToString(), UdfUtils.CreateUdfWrapper(udf), returnType).Apply0;
}
 ```

Reply:
Func<GenericRow>

Reply:
Change. Thanks for your suggestion!